### PR TITLE
feat: support vec/array impl for DnsResolver

### PR DESCRIPTION
**Pull request by NOOMA-42**
# Description
support vec/array impl for DnsResolver to enable chaining resolver together

# Related issues
https://github.com/plabayo/rama/issues/332

### CODE REVIEW

File: Cargo.lock
Comment (GlenDC) [inline Cargo.lock:1]: Can you revert this file please as it is unrelated from this PR.

File: rama-dns/src/chain.rs [hunk 1]
```diff
@@ -0,0 +1,236 @@
+use std::net::{Ipv4Addr, Ipv6Addr};
+
+use rama_net::address::Domain;
+
+use crate::{DenyAllDns, DnsResolver, InMemoryDns};
+
+#[derive(Debug)]
+pub struct DnsChainDomainResolveErr<E: 'static> {
```
Comment (GlenDC) [inline rama-dns/src/chain.rs:8]:
```suggestion
pub struct DnsChainDomainResolveErr<E> {
```

File: rama-dns/src/chain.rs [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+where
+    R: DnsResolver<Error = E> + Send,
+    E: Send + 'static,
+{
+    type Error = DnsChainDomainResolveErr<E>;
+    dns_resolver_chain_impl!();
+}
+
+#[derive(Debug)]
+enum DnsResolverType {
```
Comment (GlenDC) [inline rama-dns/src/chain.rs:76]: Why did you add this type? You can use the `Either` wrapper type for this purpose, no?

**Review by GlenDC — CHANGES_REQUESTED** [review]
Hi! thanks for picking up this PR where it was left. Left a couple of remarks already.

File: rama-dns/src/chain.rs (hunk 1 above)
Comment (NOOMA-42) [reply rama-dns/src/chain.rs:8]: Doesn't E need to be static to ensure there's no case like below ?
```rust
struct NonStaticError<'a> {
    message: &'a str  
}

let message = String::from("temp error");
let err = DnsChainDomainResolveErr {
    errors: vec![NonStaticError {
        message: &message  // The error exists after message be destroyed
    }]
};
```

Comment (GlenDC) [reply rama-dns/src/chain.rs:10]: Yes more or less, but in general it's better not to add such bounds at the definition level, but add such bounds only the impl blocks where you really need them. Otherwise you are forced to add these bounds constantly. And in some cases it's even possible that such bounds are optional and can also work in some scenarios without them.

File: rama-dns/src/chain.rs [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+
+use rama_net::address::Domain;
+
+use crate::DnsResolver;
+use rama_core::combinators::Either;
+
+/// An error that occurs when a DNS resolver chain fails to resolve a domain.
+#[derive(Debug)]
+pub struct DnsChainDomainResolveErr<E: 'static> {
+    errors: Vec<E>,
```
Comment (GlenDC) [inline rama-dns/src/chain.rs:11]: We probably want to store `BoxError` in this `Vec<..>` and do not need a generic error type at all.

File: rama-dns/src/chain.rs [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+}
+
+impl<E: std::fmt::Debug> std::fmt::Display for DnsChainDomainResolveErr<E> {
```
Comment (GlenDC) [inline rama-dns/src/chain.rs:14]: When you use BoxError internally this generic bound is no longer needed and won't exist

File: rama-dns/src/chain.rs [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(
+            f,
+            "domain resolver chain resulted in errors: {:?}",
+            self.errors
+        )
+    }
+}
+
+impl<E: std::error::Error + 'static> std::error::Error for DnsChainDomainResolveErr<E> {
```
Comment (GlenDC) [inline rama-dns/src/chain.rs:24]: same here, can be removed this bound and generic once BoxError is used

File: rama-dns/src/chain.rs [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+            }
+            Err(DnsChainDomainResolveErr { errors })
+        }
+    };
+}
+
+impl<R, E> DnsResolver for Vec<R>
+where
+    R: DnsResolver<Error = E> + Send,
+    E: Send + 'static,
```
Comment (GlenDC) [inline rama-dns/src/chain.rs:61]: this bound will change into something like `Into<BoxError>`

File: rama-dns/src/chain.rs [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+{
+    type Error = DnsChainDomainResolveErr<E>;
+
+    dns_resolver_chain_impl!();
+}
+
+impl<R, E, const N: usize> DnsResolver for [R; N]
+where
+    R: DnsResolver<Error = E> + Send,
+    E: Send + 'static,
```
Comment (GlenDC) [inline rama-dns/src/chain.rs:71]: this bound will change into something like `Into<BoxError>`

File: rama-dns/src/variant.rs [hunk 8]
```diff
@@ -0,0 +1,118 @@
+use crate::DnsResolver;
+use rama_net::address::Domain;
+use std::net::{Ipv4Addr, Ipv6Addr};
+macro_rules! impl_dns_resolver_either_either {
```
Comment (GlenDC) [inline rama-dns/src/variant.rs:4]:
```suggestion

macro_rules! impl_dns_resolver_either_either {
```

**Review by GlenDC — CHANGES_REQUESTED** [review]
Different DnsResolvers probably have their own error types. So in that case, and given this is a generic feature, best to collect the errors as `BoxError` so that we can collect the errors regardless of the types.

File: rama-dns/src/chain.rs (hunk 3 above)
Comment (NOOMA-42) [reply rama-dns/src/chain.rs]: I added Send + Sync bounds to ensure thread safe

**Comment by NOOMA-42** [comment]
> Different DnsResolvers probably have their own error types. So in that case, and given this is a generic feature, best to collect the errors as `BoxError` so that we can collect the errors regardless of the types.

I have a more general question which might outside of the scope of this pull request. When do we have to implement a specific error type and when to implement only box? We could discuss in Discord if this is inappropriate to discuss here.

File: rama-dns/src/in_memory.rs [hunk 9]
```diff
@@ -59,6 +64,16 @@ impl InMemoryDns {
         self
     }
 
+    /// Convenience method to insert an IPv4 address for a domain.
+    pub fn insert_ipv4(&mut self, name: impl Into<Domain>, addr: Ipv4Addr) -> &mut Self {
```
Comment (GlenDC) [inline rama-dns/src/in_memory.rs:68]:
```suggestion
    pub fn insert_ipv4(&mut self, name: Domain, addr: Ipv4Addr) -> &mut Self {
```

There is not a whole lot that can be turned directly into a domain as an infallible process. So might as well just expect the concrete type itself.

File: rama-dns/src/in_memory.rs [hunk 10]
```diff
@@ -59,6 +64,16 @@ impl InMemoryDns {
...
+        self.insert(name.into(), vec![IpAddr::V4(addr)])
+    }
+
+    /// Convenience method to insert an IPv6 address for a domain.
+    pub fn insert_ipv6(&mut self, name: impl Into<Domain>, addr: Ipv6Addr) -> &mut Self {
```
Comment (GlenDC) [inline rama-dns/src/in_memory.rs:73]:
```suggestion
    pub fn insert_ipv6(&mut self, name: Domain, addr: Ipv6Addr) -> &mut Self {
```

There is not a whole lot that can be turned directly into a domain as an infallible process. So might as well just expect the concrete type itself.

There is however some stuff that can be turned infallible into an Ipv4/Ipv6 addr (e.g. [u8; 4)). So might be nice addition to turn the addr part into an impl parameter for both your insert methods.

File: rama-dns/src/lib.rs [hunk 11]
```diff
@@ -129,3 +90,7 @@ pub use in_memory::{DnsOverwrite, DomainNotMappedErr, InMemoryDns};
 mod deny_all;
 #[doc(inline)]
 pub use deny_all::{DenyAllDns, DnsDeniedError};
+
+pub mod chain;
+
+pub mod variant;
```
Comment (GlenDC) [inline rama-dns/src/lib.rs:96]:
```suggestion
mod variant;
```

AFAIK no reason to have this public. It's just about implementing something.

**Review by GlenDC — CHANGES_REQUESTED** [review]
Minor changes still to go, other than that it seems this PR is about to be ready to be merged into the main branch! thx a lot!!!

File: rama-dns/src/in_memory.rs (hunk 10 above)
Comment (NOOMA-42) [reply rama-dns/src/in_memory.rs:73]: I made it a general method. let me know if this is okay

**Review by GlenDC — APPROVED** [review]

**Comment by NOOMA-42** [comment]
Thanks for this detailed explanation :) @GlenDC
